src/Orleans.Runtime/GrainDirectory/GrainDirectoryHandoffManager.cs

```diff
@@ -212,37 +208,79 @@ internal void ProcessSiloAddEvent(DirectoryMembershipSnapshot membershipSnapshot
+                    if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Splitting my partition between me and " + addedSilo);
+                    GrainDirectoryPartition splitPart = localDirectory.DirectoryPartition.Split(
+                        grain =>
+                        {
+                            var s = membershipSnapshot.CalculateGrainDirectoryPartition(grain);
+                            return (s != null) && !this.localSiloDetails.SiloAddress.Equals(s);
+                        }, false);
+                    List<ActivationAddress> splitPartListSingle = splitPart.ToListOfActivations(true);
+                    List<ActivationAddress> splitPartListMulti = splitPart.ToListOfActivations(false);
+
+                    EnqueueOperation(
+                        $"{nameof(ProcessSiloAddEvent)}({addedSilo})",
+                        () => ProcessAddedSiloAsync(membershipSnapshot, addedSilo, splitPartListSingle, splitPartListMulti));
+                }
+                else
+                {
+                    // adjust partitions by splitting them accordingly between new and old silos
+                    SiloAddress predecessorOfNewSilo = membershipSnapshot.FindPredecessor(addedSilo);
+                    if (!directoryPartitionsMap.TryGetValue(predecessorOfNewSilo, out var partitionToSplit))
+                    {
+                        // we should have the partition of the predcessor of our new successor
+                        logger.LogWarning(
+                            (int)ErrorCode.DirectoryPartitionPredecessorExpected,
+                            "This silo is expected to hold directory partition of {Predecessor}",
+                            predecessorOfNewSilo);
+                    }
+                    else
                     {
-                        var s = membershipSnapshot.CalculateGrainDirectoryPartition(grain);
-                        return (s != null) && !this.localSiloDetails.SiloAddress.Equals(s);
-                    }, false);
-                List<ActivationAddress> splitPartListSingle = splitPart.ToListOfActivations(true);
-                List<ActivationAddress> splitPartListMulti = splitPart.ToListOfActivations(false);
-
-                EnqueueOperation(
-                    $"{nameof(ProcessSiloAddEvent)}({addedSilo})",
-                    () => ProcessAddedSiloAsync(membershipSnapshot, addedSilo, splitPartListSingle, splitPartListMulti));
-            
+                        if (logger.IsEnabled(LogLevel.Debug))
+                        {
+                            logger.LogDebug("Splitting partition of {Predecessor} and creating a copy for {AddedSilo}", predecessorOfNewSilo, addedSilo);
+                        }
+
+                        GrainDirectoryPartition splitPart = partitionToSplit.Split(
+                            grain =>
+                            {
+                                // Need to review the 2nd line condition.
```

Comment:
Why it needs to be reviewed? Can the reason be added to comments succintly?

Reply:
Not sure, this code just moved from one place to another (i.e, not my code). Could probably remove it.